src/ucs/datastruct/conn_match.c

```diff
@@ -175,50 +175,99 @@ void ucs_conn_match_insert(ucs_conn_match_ctx_t *conn_match_ctx,
+                         int delete_from_queue)
 {
     char UCS_V_UNUSED address_str[UCS_CONN_MATCH_ADDRESS_STR_MAX];
-    ucs_conn_match_peer_t *peer;
     ucs_conn_match_elem_t *elem;
     ucs_hlist_head_t *head;
-    khiter_t iter;
 
-    peer = ucs_conn_match_peer_alloc(conn_match_ctx, address);
-    iter = kh_get(ucs_conn_match, &conn_match_ctx->hash, peer);
-    ucs_free(peer);
-    if (iter == kh_end(&conn_match_ctx->hash)) {
-        goto notfound; /* no hash entry */
-    }
-
-    peer = kh_key(&conn_match_ctx->hash, iter);
     head = &peer->conn_q[conn_queue_type];
 
     ucs_hlist_for_each(elem, head, list) {
         if (conn_match_ctx->ops.get_conn_sn(elem) == conn_sn) {
-            ucs_hlist_del(head, &elem->list);
+            if (delete_from_queue) {
+                ucs_hlist_del(head, &elem->list);
+            }
+
             ucs_trace("match_ctx %p: matched %s conn_match %p by address %s conn_sn %zu",
                       conn_match_ctx, ucs_conn_match_queue_title[conn_queue_type], elem,
                       conn_match_ctx->ops.address_str(conn_match_ctx,
-                                                      address, address_str,
+                                                      &peer->address, address_str,
                                                       UCS_CONN_MATCH_ADDRESS_STR_MAX),
                       conn_sn);
             return elem;
         }
     }
 
-notfound:
     ucs_trace("match_ctx %p: %s address %s conn_sn %zu not found",
               conn_match_ctx, ucs_conn_match_queue_title[conn_queue_type],
               conn_match_ctx->ops.address_str(conn_match_ctx,
-                                              address, address_str,
+                                              &peer->address, address_str,
                                               UCS_CONN_MATCH_ADDRESS_STR_MAX),
               conn_sn);
     return NULL;
 }
 
+static ucs_conn_match_peer_t *
+ucs_conn_match_find_peer(ucs_conn_match_ctx_t *conn_match_ctx,
+                         const void *address)
+{
+    char UCS_V_UNUSED address_str[UCS_CONN_MATCH_ADDRESS_STR_MAX];
+    ucs_conn_match_peer_t *peer;
+    khiter_t iter;
+
+    peer = ucs_conn_match_peer_alloc(conn_match_ctx, address);
+    iter = kh_get(ucs_conn_match, &conn_match_ctx->hash, peer);
+    ucs_free(peer);
+    if (iter == kh_end(&conn_match_ctx->hash)) {
+        /* no hash entry */
+        ucs_trace("match_ctx %p: address %s not found (no hash entry)",
+                  conn_match_ctx,
+                  conn_match_ctx->ops.address_str(conn_match_ctx,
+                                                  address, address_str,
+                                                  UCS_CONN_MATCH_ADDRESS_STR_MAX));
+        return NULL;
+    }
+
+    return kh_key(&conn_match_ctx->hash, iter);
+}
+
+ucs_conn_match_elem_t *
+ucs_conn_match_get_elem(ucs_conn_match_ctx_t *conn_match_ctx,                        
```

Comment:
do we really need to split to 2 helper functions (ucs_conn_match_find_elem and ucs_conn_match_find_peer), seems more complicated, each of them is used only once?

What is the use-case to pass UCS_CONN_MATCH_QUEUE_ANY?

Reply:
> What is the use-case to pass UCS_CONN_MATCH_QUEUE_ANY?

it will be used by UD to find the connection matching element (UD's EP) in the expected or the unexpected queue when CREQ received. It is needed since CREQ can be duplicated (resends of the CREQ packet) and it should be found in the unexpected queue in this case; in the normal case when EP with the same conn_sn is created thru APIl, it should be found in the expected queue.

> do we really need to split to 2 helper functions (ucs_conn_match_find_elem and ucs_conn_match_find_peer), seems more complicated, each of them is used only once?

fixed